Add Package Instances #2691

AyanSinhaMahapatra opened the pull request:
This PR adds the PackageInstance class and functions to group package
manifests and package data as top level package instances.

Existing package data are ported to this new approach.

Reference: https://github.com/nexB/scancode-toolkit/issues/2098 
Reference: https://github.com/nexB/scancode-toolkit/issues/2691 
Reference: https://github.com/nexB/scancode-toolkit/issues/2692
Reference: https://github.com/nexB/scancode-toolkit/issues/2693
Reference: https://github.com/nexB/scancode-toolkit/issues/2843 
Reference: https://github.com/nexB/scancode-toolkit/issues/2652 
Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

AyanSinhaMahapatra committed:
Add initial package instance creation

Adds the PackageInstance class and functions to group package manifests
and create a package instance out of them. Adds a sample example case for
python.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add function to merge package data

Adds functions to merge package data from multiple package
manifests into a package instance. Adds tests for a simple
python manifests case.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Adds PackageManifest classes for cran

Creates PackageManifest classes for cran description manifests
and overrides the methods for detection and PackageManifest creation.

See #2748

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Adds PackageManifest class for maven

Creates PackageManifest class for maven pom.xml  manifests
and overrides the methods for detection and PackageManifest creation.

See #2748

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

JonoYang commented on tests/packagedcode/test_package_instance.py:
```diff
@@ -0,0 +1,41 @@
+        result_file = self.get_temp_file('json')
+        expected_file = self.get_test_loc('instance/python-package-instance-expected.json')
+        run_scan_click(['--package', '--strip-root', '--processes', '-1', test_dir, '--json', result_file])
+        check_json_scan(expected_file, result_file, regen=True)
```
`regen` should be set to `False`

AyanSinhaMahapatra replied:
I was having some problems with the `package_uuid` field as it had a UUID that was different for each test run. So I forgot to make this False, now this is resolved. I've also modified the `check_json_scan` to accept a new argument `remove_package_uuid` which removes the UUID qualifier from the `package_uuid` pURL string.

JonoYang commented on src/packagedcode/__init__.py:
```diff
@@ -110,28 +110,36 @@
+    for cls in PACKAGE_INSTANCE_TYPES
+}
+
+def check_package_manifest_classes():
```
Is `check_package_manifest_classes()` called anywhere? Previously, the code in the function was outside of a function. My assumption was that whenever the `packagedcode` module is imported, this code that checks `PACKAGE_MANIFESTS_BY_TYPE` and `PACKAGE_MANIFEST_TYPES` would always run and give us an exception whenever a package was not properly set.

AyanSinhaMahapatra replied:
See [here](https://github.com/nexB/scancode-toolkit/pull/2748#discussion_r761111416) this was supposed to be moved into a test instead, but it was not completed. Now I've pushed the proper tests.

JonoYang commented on src/packagedcode/__init__.py:
```diff
@@ -110,28 +110,36 @@
+    cls.default_type: cls
+    for cls in PACKAGE_INSTANCE_TYPES
+}
+
```
Style nitpick: there should be an extra new line here.

AyanSinhaMahapatra replied:
Added, thanks!

AyanSinhaMahapatra replied:
Added, thanks!

JonoYang commented on src/packagedcode/models.py:
```diff
@@ -765,6 +765,251 @@ def recognize(cls, location):
+            mapping.pop(attribute)
+        
+        # Remove attributes which are BasePackage functions
+        for attribute in ('repository_homepage_url', 'repository_download_url', 'api_data_url'):
```
I have no strong feelings one way or the other about this, but I am curious as to why `repository_homepage_url`, `repository_download_url`, and `api_data_url` are removed from the package data that is to be returned.

AyanSinhaMahapatra replied:
So this is a helper function used in the merging of package_manifest data into a package_instance [here](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L897). These are basically fields which are not set as values in a package data, but obtained from functions like [this](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L320) in the `to_dict()`. So we don't need to merge these data from the mappings.

JonoYang reviewed (commented):
@AyanSinhaMahapatra Initial work on the `PackageInstance` model looks good. I just have some minor things I pointed out.

AyanSinhaMahapatra committed:
Modify package instance tests

Modifies package instance tests to remove package_uuid fields,
and adds tests for package instance/manifest type uniqueness.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add proper package_uuid

Adds package_uuid as the pURL string for the package instance,
with it's UUID as a qualifier in the pURL.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

pombredanne commented:
re: https://github.com/nexB/scancode-toolkit/pull/2825#discussion_r800437754

> I was having some problems with the package_uuid field as it had a UUID that was different for each test run. So I forgot to make this False, now this is resolved. I've also modified the check_json_scan to accept a new argument remove_package_uuid which removes the UUID qualifier from the package_uuid pURL string.


There are three ways to handle these cases:

1. ignore it as you have done but you need to have at least one test case that would not ignore the UUID to ensure this is generated correctly.
2. use mocks in tests so that you have a stable mocked UUID created and do not need to ignore it
3. add something like an extra `_test=True` argument to the tested function(s) that is called only in tests and that creates a static UUID that is well known for testing.

The more generic solution is 2. ... use mock.

AyanSinhaMahapatra committed:
Update tests and expectations for python package instance

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Activate package instance creation for others

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectations for top level packages

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove obsolete root_path

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix package instance creation bugs

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectaions for package instances

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add `for_packages` file level attribute #2693

Adds `for_packages` attribute to files to have `package_uuid` fields for the
package instance this file belongs to.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update test expectations with `for_package` #2693

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify "package_manifests" to "package_data" #2843

Modify and rename the resource-level attribute added for a package scan
to "package_data", replacing "package_manifests". Also replaces
"package_manifest_paths" in top-level "packages" to "package_data_paths".
Changes API function get_package_manifests to get_package_data.
Also renames other variables with similar naming.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add changes in test data for "package_data" renaming #2843

Regenate tests after "package_manifests" -> "package_data" renaming
and add the corresponding chages to the test files.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Add top-level `dependencies` list #2843

Adds DependencyInstance to packagedcode models.
Adds top-level dependencies list for DependencyInstance objects.
Adds DependencyInstance creation to process_codebase in packagedcode
plugin.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modifies test expectations #2843

Regenerates test expectations with top-level `dependencies` list.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify `remove_instance_uuid` option to include dependencies

Modify `remove_package_uuid` flag to `remove_instance_uuid` to
include dependencies.
Add `ignore_instance_uuid` attribute and a new test that has UUIDs
in the test file.
Add XFAIL for failing test.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into 2098-top-level-packages

AyanSinhaMahapatra committed:
Fix test failures from merging develop

Add node_modules as ignored paths for NPM package instance creation.
Regenerate test expectations after merging from develop.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix maven and nuget bugs

Adds fix for nuget and maven bugs which were in the way of activating
their respective package instance creation.

Also fixes URL creation for lockfiles.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix package checksum reporting

Package checksums were being wrongly ignored in package instance
data update from detected package data, this commit fixes that.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix dparse2 pypi detection bug

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update Package models names

Update Package->PackageData, PackageInstance->Package and
PackageData->PackageDataFile for consistency. Also rename
"package_data_paths" to "package_data_files" and "is_package_data"
to "is_package_data_file".

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove redundant dependency info from packages

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Merge branch 'develop' into 2098-top-level-packages

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update build_gradle.py to new model names

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Fix type/class ambiguity in package class lists

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Remove "root_path" package data attribute

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Update CHANGELOG with package detection changes

This updates the CHANGELOG with details about specific changes
in package detection, in the API and the data model.
Also bumps the scancode data format to 3.0.0 as there are
major data format changes.

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra committed:
Modify DependentPackage Atrributes

Rename "requirement" to "extracted_requirement" and add new attribute
"resolved_package".

Signed-off-by: Ayan Sinha Mahapatra <ayansmahapatra@gmail.com>

AyanSinhaMahapatra commented:
@pombredanne This is ready!

pombredanne reviewed (approved):
LGTM. Thanks! merging now.
